src/core/qgsvariantutils.cpp

```diff
@@ -632,3 +633,111 @@ QVariant QgsVariantUtils::createNullVariant( QMetaType::Type metaType )
 #endif
 
 }
+
+QString QgsVariantUtils::displayString( const QVariant &variant, int precision )
+{
+
+  auto _displayString = [ ]( const QVariant & variant, int precision ) -> QString
+  {
+
+    if ( QgsVariantUtils::isNull( variant ) )
+    {
+      return QgsApplication::nullRepresentation();
+    }
+
+    // Special treatment for numeric types if group separator is set or decimalPoint is not a dot
+    if ( variant.userType() == QMetaType::Type::Double )
+    {
+      // Locales with decimal point != '.' or that require group separator: use QLocale
+      if ( QLocale().decimalPoint() != '.' ||
+           !( QLocale().numberOptions() & QLocale::NumberOption::OmitGroupSeparator ) )
+      {
+        if ( precision > 0 )
+        {
+          if ( -1 < variant.toDouble() && variant.toDouble() < 1 )
+          {
+            return QLocale().toString( variant.toDouble(), 'g', precision );
+          }
+          else
+          {
+            return QLocale().toString( variant.toDouble(), 'f', precision );
+          }
+        }
+        else
+        {
+          // Precision is not set, let's guess it from the
+          // standard conversion to string
+          const QString str( variant.toString() );
+          const int dotPosition( static_cast<int>( str.indexOf( '.' ) ) );
+          int precision;
+          if ( dotPosition < 0 && str.toLower().indexOf( 'e' ) < 0 )
```

Comment:
This is more efficient:

```suggestion
          if ( dotPosition < 0 && str.indexOf( 'e', 0, Qt::CaseInsensitive ) < 0 )
```